### ISSUE TITLE
Unserialised cross-node sort key updates

### ISSUE THREAD

**Issue by domodwyer** [influxdata/influxdb_iox#6439]
## Sort Keys

A sort key defines the sort order of data in a parquet file. The ingester produces the sort key, the compactors use the sort key when compacting to optimise the data sort order for the querier, which uses the sort key during queries (chunk deduplication). We aim to avoid re-sorting data as much as possible for performance reasons.

When an ingester observes a new column, the sort key is updated according to the following logic:

https://github.com/influxdata/influxdb_iox/blob/94c2f94ea150d4ecd82e43d97ca15d2efa284f8a/schema/src/sort.rs#L478-L485

This is not commutative, requiring updates to the key be serialised, which in turn also requires persistence operations to be serialised (at least, those that update the sort key). This is handled in ingester2 by serialising persist operations per partition that [update the sort key](https://github.com/influxdata/influxdb_iox/pull/6416).

## The Problem

Previously (ingester1) all writes for a given `(namespace, table)` tuple were (effectively) consistently hashed to the same ingester instance. This meant that so long the instance serialised persist operations / sort key updates internally, updates to the sort key were serialised.

In ingester2, the current implementation of the router round-robins requests over all N routers without any consistency of data placement. This causes multiple ingester instances to need to update the sort key for the same partition, potentially at the same time:

| Ingester A| Ingester B|
|--|--|
| write A=1| |
| | write B=1|
| persist (`sort_key=A,time`)| |
| | persist (`sort_key=B,time`)|

This sort key update for B conflicts with A. This example could be mitigated as the column additions are disparate (you could merge them with "lexicographically greater wins" to consistently produce `A,B`). This doesn't hold for more complex examples, such as two instances deriving different orders for the same columns: `A,B,C` and `A,C,B`.

### Replication Complications

This also makes it difficult to support persisting from replicas should a primary be (believed to be) unavailable. It's impossible to know for sure that a primary is down, and therefore that the sort key will not be updated concurrently.

### Retries

We could read the sort key from the catalog, generate the parquet file in the ingester (sorting the data according to the sort key) and then open a transaction, check the sort key and abort if it has changed (and differs). Retry infinitely until the transaction succeeds. 

The number of retries grows super-linearly with parallelism (see https://github.com/influxdata/influxdb_iox/pull/6416) but this is exasperated across nodes. The update could race with N nodes and **may never terminate**. It is possible one or more ingesters are "starved" by faster ingesters constantly adding new sort key elements, causing the starved ingesters to retry forever. 

### Consistent Hashing

Consistently hash the writes for a `(namespace, table, partition_key)` to a single ingester2 instance. This is finer-grained than `ingester1` so should help spread hot workloads, but hot partitions will still be unevenly distributed. This also precludes allowing replicas to persist replicated data if the master is down (or rather, believed to be down but actually not).

### Convergent CRDT

Tolerate sort-key changes that eventually converge.

Use a CvRDT-based (conflict-free replicated data type) ordered list with state-based replication. This allows multiple ingesters to update the sort key, and then "merge" the results in a way that deterministically derives the outcome. State replication is cheap (small state size, only one "peer" - the catalog).

A simple example using the `A,B,C` and `A,C,B` example above would be that we tolerate `A,C,B` for one file, then converge it to `A,B,C` after. N parallel persist operations **containing overlapping, differently ordered columns** may produce N-1 inefficient parquet files before converging. More complex examples such as disparate sets of columns being merged into a total order are tolerated also.

#### Downsides

I do not know if we can tolerate sort keys changing in this manor in the querier. 

This may require compaction (re-sorting L0 files generated in parallel to match the converged sort key). 

This may also require re-compacting/sorting L1/L2 files if they were produced while the sort key was converging AND this difference effects the files. I would imagine the number of instances of this to be low however due to the delay between L0 being generated, and L1 being generated. Any recompaction work could be identified by a catalog query only.
